File: src/main/java/dev/goerner/geozen/model/collections/GeometryCollection.java
package dev.goerner.geozen.model.collections;

import dev.goerner.geozen.model.CoordinateReferenceSystem;
import dev.goerner.geozen.model.Geometry;

import java.util.List;

/**
 * A {@link GeometryCollection} is a collection of {@link Geometry Geometries}.
 */
public class GeometryCollection extends Geometry {

    private final List<Geometry> geometries;

    /**
     * Creates a new {@link GeometryCollection} with the given list of {@link Geometry Geometries} and the default WGS 84
     * {@link CoordinateReferenceSystem}.
     *
     * @param geometries The list of {@link Geometry Geometries} representing the {@link GeometryCollection}.
     */
    public GeometryCollection(List<Geometry> geometries) {
        this(geometries, CoordinateReferenceSystem.WGS_84);
    }

    /**
     * Creates a new {@link GeometryCollection} with the given list of {@link Geometry Geometries} and the given {@link CoordinateReferenceSystem}.
     *
     * @param geometries                The list of {@link Geometry Geometries} representing the {@link GeometryCollection}.
     * @param coordinateReferenceSystem The {@link CoordinateReferenceSystem} of the {@link GeometryCollection}.
     */
    public GeometryCollection(List<Geometry> geometries, CoordinateReferenceSystem coordinateReferenceSystem) {
        super(coordinateReferenceSystem);
        this.geometries = List.copyOf(geometries);
    }

    @Override
    public double getFastDistanceTo(Geometry other) {
        throw new UnsupportedOperationException("Fast distance calculation not implemented yet");
    }

    @Override
    public double getExactDistanceTo(Geometry other) {
        throw new UnsupportedOperationException("Exact distance calculation not implemented yet");
    }

    public List<Geometry> getGeometries() {
        return geometries;
    }
}

File: src/main/kotlin/dev/goerner/geozen/model/collections/GeometryCollection.kt
package dev.goerner.geozen.model.collections

import dev.goerner.geozen.calc.ApproximateDistanceCalculator
import dev.goerner.geozen.calc.PreciseDistanceCalculator
import dev.goerner.geozen.model.CoordinateReferenceSystem
import dev.goerner.geozen.model.Geometry

/**
 * A [GeometryCollection] is a collection of [Geometries][Geometry].
 */
data class GeometryCollection(
    val geometries: List<Geometry>,
    override val coordinateReferenceSystem: CoordinateReferenceSystem = CoordinateReferenceSystem.WGS_84
) : Geometry(coordinateReferenceSystem) {

    init {
        geometries.forEachIndexed { index, geometry ->
            require(geometry.coordinateReferenceSystem == coordinateReferenceSystem) {
                "The coordinate reference system of each geometry (${geometry.coordinateReferenceSystem}) " +
                        "must match the coordinate reference system of the geometry collection ($coordinateReferenceSystem), " +
                        "but geometry at index $index had a different coordinate reference system"
            }
        }
    }

    override fun fastDistanceTo(other: Geometry): Double =
        ApproximateDistanceCalculator.calculate(this, other)

    override fun exactDistanceTo(other: Geometry): Double =
        PreciseDistanceCalculator.calculate(this, other)

    override fun reprojectTo(crs: CoordinateReferenceSystem): Geometry =
        if (coordinateReferenceSystem == crs) this
        else GeometryCollection(
            geometries.map { it.reprojectTo(crs) },
            crs
        )
}
